habitat_sim/simulator.py

```diff
@@ -216,17 +225,46 @@ def __set_from_config(self, config: Configuration):
         for i in range(len(self.agents)):
             self.agents[i].controls.move_filter_fn = self.step_filter
 
-        self._default_agent = self.get_agent(config.sim_cfg.default_agent_id)
+        self._default_agent_id = config.sim_cfg.default_agent_id
 
-        agent_cfg = config.agents[config.sim_cfg.default_agent_id]
-        self._sensors = {}
-        for spec in agent_cfg.sensor_specifications:
-            self._sensors[spec.uuid] = Sensor(
-                sim=self, agent=self._default_agent, sensor_id=spec.uuid
-            )
+        self.__sensors: List[Dict[str, Sensor]] = [
+            dict() for i in range(len(config.agents))
+        ]
+        self.__last_state = [None] * len(self.__sensors)
+        for agent_id, agent_cfg in enumerate(config.agents):
+            for spec in agent_cfg.sensor_specifications:
+                self._update_simulator_sensors(spec.uuid, agent_id=agent_id)
+            self.initialize_agent(agent_id)
+
+    def _update_simulator_sensors(self, uuid: str, agent_id: int = 0) -> None:
+        self.__sensors[agent_id][uuid] = Sensor(
+            sim=self, agent=self.get_agent(agent_id), sensor_id=uuid
+        )
 
-        for i in range(len(self.agents)):
-            self.initialize_agent(i)
+    def add_sensor(self, sensor_spec: SensorSpec, agent_id: int = 0) -> None:
+        if (
```

Comment:
@bigbike This is going to get messy as we add more SensorTypes. We should consider a way of making this cleaner.

Reply:
Maybe, those checks should live close to Sensor code of particular type and should be checked before particular sensor type creation in `update_simulator_sensors` method.

Reply:
@mathfac They should be handled in the C++ API or have a boolean on the Sensors that correspond which flags are valid / needed for an arbitrary sensor type.